artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/AMQPRemoteControlsSource.java

```diff
@@ -0,0 +1,238 @@
+         symbolObjectMap.put(INTERNAL_ID, message.getMessageID());
+         ref.setProtocolData(deliveryAnnotations);
+
+         refs.add(ref);
+         message.usageUp();
+      } catch (Throwable e) {
+         logger.warn(e.getMessage(), e);
+      }
+   }
+
+   @Override
+   public void postAcknowledge(MessageReference ref, AckReason reason) throws Exception {
+      if (acks && !ref.getQueue().isRemoteControl()) { // we don't call postACK on snfqueues, otherwise we would get infinite loop because of this feedback
+         Message message = createMessage(ref.getQueue().getAddress(), ref.getQueue().getName(), POST_ACK, ref.getMessage().getMessageID());
+         route(server, message);
+         ref.getMessage().usageDown();
+      }
+   }
+
+   private Message createMessage(SimpleString address, SimpleString queue, Object event, Object body) {
+      return createMessage(snfQueue.getAddress().toString(), address, queue, event, body);
+   }
+
+   /** This method is open to make it testable,
+    * do not use on your applications. */
+   public static Message createMessage(String to, SimpleString address, SimpleString queue, Object event, Object body) {
+      Header header = new Header();
+      header.setDurable(true);
+
+      Map<Symbol, Object> annotations = new HashMap<>();
+      annotations.put(EVENT_TYPE, event);
+      if (address != null) {
+         annotations.put(ADDRESS, address.toString());
+      }
+      if (queue != null) {
+         annotations.put(QUEUE, queue.toString());
+      }
+      MessageAnnotations messageAnnotations = new MessageAnnotations(annotations);
+
+      Properties properties = new Properties();
+      properties.setTo(to);
+
+      Section sectionBody = body != null ? new AmqpValue(body) : null;
+
+      ByteBuf buffer = PooledByteBufAllocator.DEFAULT.heapBuffer(1024);
+
+      try {
+         EncoderImpl encoder = TLSEncode.getEncoder();
+         encoder.setByteBuffer(new NettyWritable(buffer));
+         encoder.writeObject(header);
+         encoder.writeObject(messageAnnotations);
+         encoder.writeObject(properties);
+         if (sectionBody != null) {
+            encoder.writeObject(sectionBody);
+         }
+
+         byte[] data = new byte[buffer.writerIndex()];
+         buffer.readBytes(data);
+
+         AMQPMessage amqpMessage = new AMQPStandardMessage(0, data, null);
+         return amqpMessage;
+
+      } finally {
+         TLSEncode.getEncoder().setByteBuffer((WritableBuffer) null);
+         buffer.release();
+      }
+   }
+
+   public static void route(ActiveMQServer server, Message message) throws Exception {
+      message.setMessageID(server.getStorageManager().generateID());
+      RemoteControlRouting ctx = remoteControlRouting.get();
+      ctx.clear();
+      server.getPostOffice().route(message, ctx, false);
+   }
+
+   @Override
+   public void routingDone(List<MessageReference> refs, boolean direct) {
+      /*for (MessageReference ref : refs) {
```

Comment:
commented code

Reply:
I rebased my code...

There are two separate commits to be squashed.. one containing the rename of amp-connections 
and one to rename remote control...


I have removed routingDone from MirrorController as it was not used after all.


I'm not working tomorrow as I'm taking holiday, but I should finish the documentation early next week...


I also still need to use an PrimitiveLongHashMap instead of a HashMap on the Queue... it will be a separate commit to be squashed here...

Reply:
@michaelandrepearce / @michaelpearce-gain